### ISSUE TITLE
IPO loss resulting in NaN

### ISSUE THREAD

**Issue by orendar** [huggingface/trl#1033]
Hey, thank you for adding IPO loss to the lib!

I have been experimenting with TRL and DPO using the alignment-handbook repository and if I use 'dpo'/'hinge' losses training goes well, but if I use 'ipo' then I get NaN in both training and evaluation losses. Is there anything else I need to add if I want to experiment with IPO loss? Or should I tweak beta or any other parameter?

Thanks!

**Comment by kashif** [comment]
so the `beta` for IPO I believe needs to be different from the `beta` for DPO, can you try with an order of mag. larger/smaller beta (but within the range of the paper)

**Comment by Georgefwt** [comment]
Hello @orendar , I also encountered the same problem. May I ask how you finally resolved it?

**Comment by SaeedNajafi** [comment]
```
if self.loss_type == "ipo" or self.average_logps == "yes":
            all_logps = all_logps / loss_mask.sum(-1)
```

This if when enabled, causes nan logits and non loss.

**Comment by ductri** [comment]
Agree with @SaeedNajafi , that is the issue. 
When you do padding using max length and encounter an input with the length longer or equal to the max length, the `loss_mask` for that input would be all 0, resulting a 0 in `loss_mask.sum(-1)`.

And also, this won't happen with other losses. I don't know for whatever reason, IPO has this special pre-processing step!

For now, I'm using 
```
all_logps = all_logps / torch.clamp(loss_mask.sum(-1), 1)
```

**Comment by SaeedNajafi** [comment]
> Agree with [@SaeedNajafi](https://github.com/SaeedNajafi) , that is the issue. When you do padding using max length and encounter an input with the length longer or equal to the max length, the `loss_mask` for that input would be all 0, resulting a 0 in `loss_mask.sum(-1)`.
> 
> And also, this won't happen with other losses. I don't know for whatever reason, IPO has this special pre-processing step!
> 
> For now, I'm using
> 
> ```
> all_logps = all_logps / torch.clamp(loss_mask.sum(-1), 1)
> ```
> 
> [@orendar](https://github.com/orendar) could you re-open this issue and so someone can notice and fix this simple bug?

Totally correct, I also noticed some lengths are zero for some of the examples in the batch based on loss_mask.sum(-1).

**Comment by Georgefwt** [comment]
I figured it out — did you set `max_length` to a value ≤ 512 in `CPOConfig`? The default value of `max_prompt_length` in `CPOConfig` is 512, which would cause `self.max_length - self.max_prompt_length` to be 0 in `tokenize_row` (see [this line](https://github.com/huggingface/trl/blob/eab175d434b9bb9badee20335c7945991a26dfac/trl/trainer/cpo_trainer.py#L522C1-L523C1)).

However, it seems this issue has now been addressed, since the current version of TRL includes an assertion here: [assertion line](https://github.com/huggingface/trl/blob/eab175d434b9bb9badee20335c7945991a26dfac/trl/trainer/cpo_trainer.py#L245C1-L246C1). So it looks like the problem has been resolved?
